protocols/consensus-gossip/tuple-space.md

```diff
@@ -0,0 +1,192 @@
+
+[^todo1]: Formalize using [Making CRDTs Byzantine Fault Tolerant](https://martin.kleppmann.com/papers/bft-crdt-papoc22.pdf) as basis; "Many CRDTs, such as Logoot [44] and Treedoc [ 36], assign a unique identifier to each item (e.g. each element in a sequence); the data structure does not allow multiple items with the same ID, since then an ID would be ambiguous.”
+
+[^todo2]: This should be trivial from the fact that the first approach is essentially Tendermint using Gossip Communication.
+
+
+### Tuple Removal and Garbage Collection
+
+In both approaches for synchronization, the tuple space could grow indefinitely, given that the number of heights and rounds is infinite.
+To save memory, entries should be removed from the tuple space as soon as they become stale, that is, they are no longer useful.
+For example, if a new height is started, all entries corresponding toe previous heights become stale.
+
+In general, simply forgetting stale entries in the local view would save the most space.
+However, it could lead to entries being added back or circulating on the network.
+Even if re-adding tuples does not compromise CometBFT's correctness, removed entries should not be added again, for performance and resource utilization sake.
+
+Re-adding entries may be prevented by keeping _tombstones_ for the removed entries.
+We note the tombstone for an entry $e$ as $\bar{e}$.
+
+With time, even small tombstones may accrue and need to be garbage collected, in which cause the corresponding entry may be added again; again, this will not break correctness and as long as tombstones are kept for long enough, the risk of re-adding is minimal.
+
+In the case of the Tendermint algorithm we note that staleness comes from adding newer messages (belonging to higher rounds and heights) to the tuple space.
+Hence, in Approach Two, if as an optimization these newer messages are exchanged first, then the stale messages can be excluded before being shared to other nodes that might have forgotten them and tombstones may not be needed at all.
+
+
+### Querying the Tuple Space
+
+The tuple space is consulted through queries, which have the same form as the entries.
+Queries return all entries whose values match those in the query, where a `*` matches all values.
+For example, suppose a node's local view of the tuple space has the following entries, here organized as rows of a table for easier visualization:
+
+| Height | Round | Step     | Validator | Value |
+|--------|-------|----------|-----------|-------|
+| 1      | 0     | Proposal | p         | pval  |
+| 1      | 0     | PreVote  | p         | vval  |
+| 1      | 1     | PreCommit| q         | cval  |
+| 2      | 0     | Proposal | p         | pval' |
+| 2      | 2     | PreVote  | q         | vval' |
+| 2      | 3     | PreCommit| q         | cval' |
+
+- Query $\lang 0, 0, Proposal, p, * \rang$ returns $\{ \lang 0, 0, Proposal, p, pval \rang \}$
```

Comment:
Here we could say if `p` is faulty, query can give any result (allowing for two-faces perceptions of Byzantine faults)